### DISCUSSION TITLE
New webpart integration

### DISCUSSION THREAD

**Discussion by zaiby** [microsoft-search/pnp-modern-search#2457] [Q&A]
Hi,

One of our clients uses the pnp-modern-search, and would like to have some extra features which are not available as far as I know.
They use external connector, and would like to have features like, dynamic sorting, clear all filters button, and they have found a 'bug' which can only be fixed by setting a language header in the query call.

Initially I was going to create a separate webpart for this, but can see that all webparts are bundled together, so communication is going to be an issue between the webparts.

So my question is: Do you support external webparts / or have a plan to do so?
Or should I just make the changes in the main code (Add new webpart in this solution), and hope for it to be approved, and if not approved, the customer will be forced to use a specific version.

Regards Zuhaib Mian

**Comment by FranckyC** [comment]
Hi @zaiby, what Web Part you are referring to? Web Parts in the solution are connected through SPFx dynamic data so you can totally create your own reproducing the expected schema and connect with the default ones.

For other enhancements, feel free to PR. If they meet the general interest and follow the solution principles and guidelines, they will be likely approved. However, I can already tell you we won't add an other Web Part to the solution for sure.

**Comment by zaiby** [comment]
Hey, ah okay, was not sure if they used 'dynamic data' or not, but that sounds great.
So most of the customization can be made in a separate webpart then, the only thing left, is to dynamically set language header, do you know if that is possible? Or can you pre-approve that logic, so I can do the change in the main solution.

**Reply by FranckyC** [reply]
What do you mean by language header? What is the issue right now and how do you plan ti resolve it?

**Reply by zaiby** [reply]
We have an issue, that if a document is lets say is in Chinese (but all metadata/properties are in English), then we can't search it by a given property the user normally uses to search for documents.
So I have figured out, if I then can set the language header to cz i think it was, then the document is returned fine.
This happens on some documents not all, and the issue is also there for other languages. 

and language header for the query fetch: Accept-Language

And solution is to give the users a dropdown with language names, so they can select cz if they think/know it was a Chinese doc.

If the language header could be specified using 'dynamic data' would also work :)

**Reply by zaiby** [reply]
@FranckyC Hi, did you have a look at this?

I've solved most of my other features using custom-web-component and then triggering events, so this last bug/issue could also be solved in similar manner if approved, that we can trigger an event from web-comp, which sets a language header, or can support multiple headers?

**Reply by wobba** [reply]
By default PnP set the language based on the SharePoint UX language, or the language of a translated modern page. I see we don't set any specific language value on the Graph data source today, so this should likely be changed. I don't see any public docs on what to set and will ask the Graph team.

**Reply by zaiby** [reply]
Hi,

So when PnP sets this language, it would be very helpful if it can also be set/changed by the user/customization.

As to fixing this issue the language has to match the document language (The document user is searching for) otherwise no result will appear. In old/on-prem it was possible to provide culture code, in the query, I need an option like that, or that microsoft fixes this bug so the language info is not required for some documents.

**Reply by wobba** [reply]
I agree some users might want to set the language, but most user won't. They expect it to just work. I have enquired as to how to pass in culture for the Graph API.

**Reply by zaiby** [reply]
thanks, please keep me updated :)

**Reply by wobba** [reply]
We'll add the change to the pnp web parts to set this header based on UX language for the Graph Provider.

**Reply by zaiby** [reply]
And will you allow this to be set by the user/dynamically ?

**Reply by wobba** [reply]
Dynamically - same as we do for the SharePoint provider. Unless someone PR's to add the option as web-part setting override. If the user is to pick that is a new webcomponent or similar.

**Reply by wobba** [reply]
@zaiby fixed in the dev branch now.

**Reply by zaiby** [reply]
Great, will check it out :)

**Reply by wobba** [reply]
..and in the 4.8 release :)
